New BTD class layout and member functions - Part 1

RevathiJambunathan opened the pull request:
This is the first PR in the series "reformat back-transformed Diagnostics" 

This PR includes the following changes to diagnostics.

- Created a new Back-transformed diagnostics class::BTDiagnostics in Source/Diagnostics/BTDiagnostics.H 

- A new functor "class::BackTransformFunctor" is included in Source/Diagnostics/ComputeDiagFunctor/BackTransformFunctor.H. The main objective of this functor is to compute the core part of BTD implementation, where, cell-centered data is sliced for a given snapshot and back-transformed to lab-frame.

- A lot of the implementation in the Diagnostics class was specific to FullDiagnostics. So I have moved functions relevant only to FullDiagnostics to class::FullDiagnostics.

- Now, the layout looks like this :
  * Base class::Diagnostics
  * Two derived classes: class::FullDiangotics and class::BTDiagnostics
  * Member variables, such as, vector of multifabs storing output buffers on multiple levels, which are common to both derived classes are in the Diagnostics class.
   * Similarly, Functions common to both dervied classes are also in the Base class.
   * In the new BTDiagnostics class, member variables required to construct the class and build the back-transformed data is included.
   * The functions to initialize these vectors/varaibles are currently empty with comments describing the main objective of the that function or steps to be included in the function. (These functions will be "filled" with implementation in subsequent PRs)

addition to the input file to perform Back-Transformed diagnostics

```
diagnostics.diags_names = diag1 diag2

diag1.period = 2
diag1.diag_type = Full

diag2.diag_type = BackTransformed
diag2.do_back_transformed_fields = 1
diag2.num_snapshots_lab = 4
diag2.dz_snapshots_lab = 0.001
diag2.fields_to_plot = Ex Ey Ez By rho
```

MaxThevenet commented on Source/Diagnostics/Diagnostics.H:
```diff
@@ -18,99 +18,84 @@ class Diagnostics
-    void Flush ();
-    /** Flush raw data */
-    void FlushRaw ();
+    virtual void Flush (){};
```
```suggestion
    /** \brief Flush particle and field buffers to file using the FlushFormat member variable.
     *
     * This function should belong to class Diagnostics and not be virtual, as it flushes
     * the particle buffers (name not found yet) and the field buffers (m_mf_output), both
     * of which are members of Diagnostics. Yet, the implementation is left to derived classes
     * for now because:
     * - The functions underlying FlushFormat::WriteToFile expect a geometry object, which is
     *   WarpX::geom for FullDiagnostics but should be constructed for BTDiagnostics;
     * - The functions underlying FlushFormat::WriteToFile do not support writing a buffer to file
     *   multiple times yet.
     * When these are fixed, the implementation of Flush should be in Diagnostics.cpp
     */
    virtual void Flush (){};
```

MaxThevenet commented on Source/Diagnostics/Diagnostics.H:
```diff
@@ -18,99 +18,84 @@ class Diagnostics
-      * \param[in] lev level on which the vector of unique_ptrs to field functors is initialized.
-      */
-    void InitializeFieldFunctors (int lev);
+     * \param[in] lev level on which the vector of unique_ptrs to field functors is initialized.
+     */
+    virtual void InitializeFieldFunctors (int lev){};
```
Ideally this function would be pure virtual because derived classes MUST implement it, so let's see if this is doable.
```suggestion
     *
     * Derived classes MUST implement this function, and it must allocate m_all_field_functors
     * and fill it with ComputeDiagFunctor objects.
     * \param[in] lev level on which the vector of unique_ptrs to field functors is initialized.
     */
    virtual void InitializeFieldFunctors (int lev) = 0;
```
An alternative is to implement it in Diagnostics directly, and have it call virtual functions (i.e., merge FullDiagnostics::InitializeFieldFunctors and BTDiagnostics::InitializeFieldFunctors).
The same applies for InitData

RevathiJambunathan replied:
the size of m_field_functors = m_varnames.size() for FullDiagnostics and =1 for BTDiagnostics.
Since it cannot be merged, I will make this a virtual function.

MaxThevenet commented on Source/Diagnostics/Diagnostics.H:
```diff
@@ -18,99 +18,84 @@ class Diagnostics
+    virtual void FlushRaw (){};
     /** Initialize pointers to main fields and allocate output multifab m_mf_output. */
-    void InitData ();
+    virtual void InitData (){};
```
see below `InitializeFieldFunctors` 
```suggestion
    virtual void InitData () = 0;
```

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.cpp:
```diff
@@ -0,0 +1,203 @@
+bool
+BTDiagnostics::DoDump (int step, bool force_flush)
+{
+    // always set to true for BTDiagnostics since back-transform buffers are potentially
+    // filled every timstep. They will be flushed if m_buffer_counter for
+    // the snapshot is equal to m_buffer_size
+    return true;
```
I think the spirit of this function is more something like
```c++
BTDiagnostics::DoDump(buffer)
if (buffer.is_full()) || (force_flush){
    return true;
}
return false;
```

RevathiJambunathan replied:
done.

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -0,0 +1,152 @@
+#define WARPX_BTDIAGNOSTICS_H_
+
+#include "Diagnostics.H"
+using namespace amrex::literals;
```
We don't want to have any `using namespace` in a .H file. I think the reason you add it here is for the `_rt`, but this is mostly important inside kernels, I don't think you need it here. If you really want to keep it, I would suggest putting the body of function `dz_lab` in the cpp file instead, as it is the only one where you need `_rt`.
```suggestion
```

MaxThevenet commented on Source/Diagnostics/BTDiagnostics.cpp:
```diff
@@ -0,0 +1,203 @@
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(snapshot_interval_is_specified,
+        "For back-transformed diagnostics, user should specify either dz_snapshots_lab or dt_snapshots_lab");
+
+    m_map_actual_fields_to_dump.resize(m_varnames.size());
+    for (int i=0; i < m_varnames.size(); ++i) {
+        m_map_actual_fields_to_dump[i] = m_possible_fields_to_dump[ m_varnames[i] ];
+    }
```
This handling already exists in FullDiagnostics. Shouldn't it be common, and implemented in class `Diagnostics`?

RevathiJambunathan replied:
In BTD, this Gpu::Vector is generated and used in the gpu-kernels when adding the transformed data to the output buffer.
Currently, LabFrameSnapShot::AddDataToBuffer(...)
access the i^th component of the multifab using the vector "m_map_actual_fields_to_dump".

Thanks for bringing up this point. May be the real use of this vector may not be in BTDiagnostics.cpp, but it would be required in BackTransformFunctor.cpp when slicing+Lorentz_transforming+copying_data (on the gpu)

RevathiJambunathan replied:
Thanks for the offline discussion. Removed these arrays since the new machinery ensures that the cc_mf.ncomp == m_mf_output.ncomp

RevathiJambunathan commented on Source/Diagnostics/Diagnostics.H:
```diff
@@ -18,99 +18,84 @@ class Diagnostics
-    /** Flush raw data */
-    void FlushRaw ();
+    virtual void Flush (){};
+    /** Flush raw data (applies only to full diagnostics)*/
```
Makes sense! Done!

RevathiJambunathan commented on Source/WarpX.cpp:
```diff
@@ -416,7 +416,7 @@ WarpX::ReadParameters ()
                 "When using back-transformed diagnostics, user should specify either dz_snapshots_lab or dt_snapshots_lab.");
 
-            pp.get("gamma_boost", gamma_boost);
+            pp.get("gamma_boost", gamma_boost); // this is read in ReadBoostedFrameParameters()
```
Thank you. forgot to remove this comment. It was written when I was trying to understand where the boosted-frame data is read. 
gamma_boost is parsed using pp.get() in WarpX.cpp 
and it is also queried in WarpXUtil.cpp instead of using the already parsed gamma_boost value. This is not necessarily a problem. Just wanted to make a note to myself to see whether the query is really necessary?

RevathiJambunathan commented:
@MaxThevenet Thank you so much for the suggestions. I have addressed them now.
1. InitData() in class::Diagnostics
2. removed the redundant (and inconsistent) vector of maps for fields
3. Moved FilterComputeAndPack() to class::Diagnostics, since Flush requires a loop for m_num_buffers.

MaxThevenet commented on Source/Diagnostics/ComputeDiagFunctors/BackTransformFunctor.cpp:
```diff
@@ -0,0 +1,8 @@
+BackTransformFunctor::BackTransformFunctor(amrex::MultiFab const * mf_src, int lev,
+                                           int ncomp, amrex::IntVect crse_ratio)
+    : ComputeDiagFunctor(ncomp, crse_ratio), m_mf_src(mf_src), m_lev(lev)
+{}
```
```suggestion
{
    // Get the right slice of each field in the CC MultiFab, transform it and store it in the output.
}
```

MaxThevenet commented on Source/Diagnostics/ComputeDiagFunctors/BackTransformFunctor.cpp:
```diff
@@ -0,0 +1,8 @@
+#include "BackTransformFunctor.H"
+
+using namespace amrex;
+
```
We stopped doing that.
```suggestion
```

RevathiJambunathan commented on Source/Diagnostics/BTDiagnostics.H:
```diff
@@ -0,0 +1,144 @@
+     *  \param[in] dt in boosted-frame at level, lev
+     *  \param[in] ref_ratio, refinement ratio at level, lev
+     */
+    amrex::Real dz_lab (amrex::Real dt, amrex::Real ref_ratio) {
+        using namespace amrex::literals;
+        return PhysConst::c * dt * 1._rt/m_beta_boost * 1._rt/m_gamma_boost * 1._rt/ref_ratio;
+    }
```
moving this to cpp.

ax3l commented on Source/Diagnostics/BTDiagnostics.cpp:
```diff
@@ -0,0 +1,210 @@
+    m_num_buffers = m_num_snapshots_lab;
+
+    // Read either dz_snapshots_lab or dt_snapshots_lab
+    bool snapshot_interval_is_specified = 0;
```
```suggestion
    bool snapshot_interval_is_specified = false;
```

ax3l replied:
```suggestion
    bool snapshot_interval_is_specified = false;
```

ax3l commented on Source/Diagnostics/BTDiagnostics.cpp:
```diff
@@ -0,0 +1,210 @@
+    snapshot_interval_is_specified = pp.query("dt_snapshots_lab", m_dt_snapshots_lab);
+    if ( pp.query("dz_snapshots_lab", m_dz_snapshots_lab) ) {
+        m_dt_snapshots_lab = m_dz_snapshots_lab/PhysConst::c;
+        snapshot_interval_is_specified = 1;
```
```suggestion
        snapshot_interval_is_specified = true;
```

ax3l commented on Source/Diagnostics/Diagnostics.H:
```diff
@@ -18,99 +18,125 @@ class Diagnostics
+    virtual ~Diagnostics () ;
     /** Pack (stack) all fields in the cell-centered output MultiFab m_mf_output.
-     * Fields are computed (e.g., cell-centered) on-the-fly using a functor.*/
+     * Fields are computed (e.g., cell-centered or back-transformed)
+       on-the-fly using a functor.*/
```
You can separate title from description in doxygen syntax like this:
```suggestion
     * 
     * Fields are computed (e.g., cell-centered or back-transformed)
       on-the-fly using a functor. */
```

ax3l commented on Source/Diagnostics/Diagnostics.H:
```diff
@@ -18,99 +18,125 @@ class Diagnostics
-    bool m_plot_raw_F = false;
-    /** Whether to dump the RZ modes */
-    bool m_dump_rz_modes = false;
+    /** format for output files, "plotfile" or "openpmd" */
```
Is this more yet? I.e. `ascent` and `sensei`, too?
```suggestion
    /** format for output files, "plotfile" or "openpmd" */
```

RevathiJambunathan replied:
Thank you for noticing this. I will add ascent and sensei too. Currently BTD does not support plotfile/openpmd/ascent/sensei and so I am adding Asserts for these formats in BTDiagnostics.cpp

ax3l commented on Source/Diagnostics/MultiDiagnostics.H:
```diff
@@ -2,9 +2,10 @@
 
 /** All types of diagnostics. */
-enum struct DiagTypes {Full};
+enum struct DiagTypes {Full, BackTransformed};
```
Does this only include Full diagnostics and derived? Just wondering about reduced diagnostics.

We can potentially also unify that enum with the string mapping in `m_format` members (e.g. we add helper functions or members for enum <-> string here).

RevathiJambunathan replied:
We discussed this offline. Full and BT diagnostics are handled independently by the object 'multi_diags' compared to ReducedDiags 'reduced_diags'. So for now, we are not merging them.

ax3l commented:
Talked to @atmyers and he'll take the in-depth look at this :)

atmyers commented:
Thanks for taking this on - the basic structure you have here looks good to me.

RevathiJambunathan commented:
Thank you for the review and suggestions @MaxThevenet @ax3l and @atmyers !
